### PR TITLE
[FLINK-33905][core] Unify the Provision of Diverse Metadata for Context-like APIs

### PR DESCRIPTION
**Pull request by WencongLiu**
## What is the purpose of the change

*Unify the provision of diverse metadata for context-like APIs.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-core/src/main/java/org/apache/flink/api/common/JobInfoImpl.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+ */
+
+package org.apache.flink.api.common;
+
+import org.apache.flink.annotation.PublicEvolving;
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+
+/** The default implementation of {@link JobInfo}. */
+@PublicEvolving
```
Comment (xintongsong) [inline flink-core/src/main/java/org/apache/flink/api/common/JobInfoImpl.java:26]: The implementation class should be `@Internal`.

File: flink-core/src/main/java/org/apache/flink/api/common/TaskInfoImpl.java [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+package org.apache.flink.api.common;
+
+import org.apache.flink.annotation.PublicEvolving;
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/** The default implementation of {@link TaskInfo}. */
+@PublicEvolving
```
Comment (xintongsong) [inline flink-core/src/main/java/org/apache/flink/api/common/TaskInfoImpl.java:27]: The implementation class should be `@Internal`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java [hunk 3]
```diff
@@ -166,7 +169,10 @@ public class Task
     /** ID which identifies the slot in which the task is supposed to run. */
     private final AllocationID allocationId;
 
-    /** TaskInfo object for this task. */
+    /** The meta information of current job. */
+    private final JobInfo jobInfo;
```
Comment (xintongsong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java:173]: Nit: This seems in the wrong commit.

File: flink-core/src/main/java/org/apache/flink/api/common/functions/RuntimeContext.java [hunk 4]
```diff
@@ -473,4 +520,26 @@ <IN, ACC, OUT> AggregatingState<IN, OUT> getAggregatingState(
...
+    /**
+     * Get the meta information of current task.
+     *
+     * @return the task meta information.
+     */
+    @PublicEvolving
+    @Nullable
+    default TaskInfo getTaskInfo() {
+        return null;
+    }
```
Comment (xintongsong) [inline flink-core/src/main/java/org/apache/flink/api/common/functions/RuntimeContext.java:544]: Why provide default implementation for these methods? And why these are `@Nullable`? What does it mean when jobinfo / taskinfo is null?

File: flink-core/src/main/java/org/apache/flink/api/common/functions/RuntimeContext.java [hunk 5]
```diff
@@ -63,14 +67,27 @@ public interface RuntimeContext {
...
      * returned ID should NOT be used for any job management tasks.
+     *
+     * @deprecated This method is deprecated since Flink 1.19. All metadata about the job should be
+     *     provided uniformly by {@link #getJobInfo()}.
+     * @see <a
+     *     href="https://cwiki.apache.org/confluence/display/FLINK/FLIP-382%3A+Unify+the+Provision+of+Diverse+Metadata+for+Context-like+APIs">
+     *     FLIP-382: Unify the Provision of Diverse Metadata for Context-like APIs </a>
      */
+    @Deprecated
     JobID getJobId();
```
Comment (xintongsong) [inline flink-core/src/main/java/org/apache/flink/api/common/functions/RuntimeContext.java:78]: Actually, we can provide default implementations for these deprecated methods. E.g., `JobID getJobId() { return getJobInfo().getJobId(); }`, so that we can remove the duplicated method implementations for the concrete classes.

File: flink-core/src/test/java/org/apache/flink/api/common/operators/base/FlatMapOperatorCollectionTest.java [hunk 6]
```diff
@@ -42,6 +43,7 @@
 import java.util.List;
 import java.util.concurrent.Future;
 
+/** The test for flat map operator. */
```
Comment (xintongsong) [inline flink-core/src/test/java/org/apache/flink/api/common/operators/base/FlatMapOperatorCollectionTest.java:46]: Such code-style clean-ups can be placed in a separated hotfix commit.

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/operator/CepRuntimeContextTest.java [hunk 7]
```diff
@@ -110,38 +111,41 @@ public void testCepRuntimeContext() {
         final String taskName = "foobarTask";
         final OperatorMetricGroup metricGroup =
                 UnregisteredMetricsGroup.createOperatorMetricGroup();
-        final int numberOfParallelSubtasks = 42;
-        final int indexOfSubtask = 43;
+        final int numberOfParallelSubtasks = 43;
+        final int indexOfSubtask = 42;
         final int attemptNumber = 1337;
-        final String taskNameWithSubtask = "barfoo";
+        final String taskNameWithSubtask = "foobarTask (43/43)#1337";
```
Comment (xintongsong) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/operator/CepRuntimeContextTest.java:117]: How are these changes related to this commit?

File: flink-connectors/flink-connector-files/src/test/java/org/apache/flink/connector/file/table/batch/compact/BatchPartitionCommitterSinkTest.java [hunk 8]
```diff
@@ -127,7 +127,7 @@ public Long timestamp() {
             };
 
     private static RuntimeContext getMockRuntimeContext() {
-        return new MockStreamingRuntimeContext(false, 0, 0) {
+        return new MockStreamingRuntimeContext(false, 1, 0) {
```
Comment (xintongsong) [inline flink-connectors/flink-connector-files/src/test/java/org/apache/flink/connector/file/table/batch/compact/BatchPartitionCommitterSinkTest.java:130]: Unrelated change

File: flink-runtime/src/main/java/org/apache/flink/runtime/execution/Environment.java [hunk 9]
```diff
@@ -119,7 +120,14 @@ public interface Environment {
     Configuration getJobConfiguration();
 
     /**
-     * Returns the {@link TaskInfo} object associated with this subtask
+     * Returns the {@link JobInfo} object associated with current job.
+     *
+     * @return JobInfo for current job
+     */
+    JobInfo getJobInfo();
```
Comment (xintongsong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/execution/Environment.java:127]: This does not match the commit message "Unify the provision of metadata in RuntimeContext"

File: flink-core/src/main/java/org/apache/flink/api/connector/source/SourceReaderContext.java [hunk 10]
```diff
@@ -39,7 +41,16 @@ public interface SourceReaderContext {
...
+     * Get the index of this subtask.
+     *
+     * @deprecated This method is deprecated since Flink 1.19. All metadata about the task should be
+     *     provided uniformly by {@link #getTaskInfo()}.
+     * @see <a
+     *     href="https://cwiki.apache.org/confluence/display/FLINK/FLIP-382%3A+Unify+the+Provision+of+Diverse+Metadata+for+Context-like+APIs">
+     *     FLIP-382: Unify the Provision of Diverse Metadata for Context-like APIs </a>
+     */
+    @Deprecated
     int getIndexOfSubtask();
```
Comment (xintongsong) [inline flink-core/src/main/java/org/apache/flink/api/connector/source/SourceReaderContext.java:54]: Same here for the deprecated methods.

File: flink-core/src/main/java/org/apache/flink/api/connector/source/SourceReaderContext.java [hunk 11]
```diff
@@ -68,8 +79,24 @@ public interface SourceReaderContext {
...
     }
+
+    /**
+     * Get the meta information of current task.
+     *
+     * @return the task meta information.
+     */
+    @PublicEvolving
+    default TaskInfo getTaskInfo() {
+        return null;
```
Comment (xintongsong) [inline flink-core/src/main/java/org/apache/flink/api/connector/source/SourceReaderContext.java:100]: Same here for the default implementation and Nullable.

File: flink-core/src/main/java/org/apache/flink/api/connector/sink2/InitContext.java [hunk 12]
```diff
@@ -57,6 +83,34 @@ public interface InitContext {
...
+    /**
+     * Get the meta information of current task.
+     *
+     * @return the task meta information.
+     */
+    @PublicEvolving
+    @Nullable
+    default TaskInfo getTaskInfo() {
+        return null;
+    }
```
Comment (xintongsong) [inline flink-core/src/main/java/org/apache/flink/api/connector/sink2/InitContext.java:115]: Same here for default and nullable.

File: flink-core/src/main/java/org/apache/flink/api/connector/sink2/InitContext.java [hunk 13]
```diff
@@ -35,16 +40,37 @@ public interface InitContext {
...
+    /**
+     * Get the id of task where the committer is running.
+     *
+     * @deprecated This method is deprecated since Flink 1.19. All metadata about the task should be
+     *     provided uniformly by {@link #getTaskInfo()}.
+     * @see <a
+     *     href="https://cwiki.apache.org/confluence/display/FLINK/FLIP-382%3A+Unify+the+Provision+of+Diverse+Metadata+for+Context-like+APIs">
+     *     FLIP-382: Unify the Provision of Diverse Metadata for Context-like APIs </a>
+     */
     int getSubtaskId();
```
Comment (xintongsong) [inline flink-core/src/main/java/org/apache/flink/api/connector/sink2/InitContext.java:52]: 1. Aren't we deprecating these methods?
2. Same for the default implementation.

File: flink-core/src/main/java/org/apache/flink/core/failure/FailureEnricher.java [hunk 14]
```diff
@@ -84,14 +88,26 @@ enum FailureType {
          * Get the ID of the job.
          *
          * @return the ID of the job
+         * @deprecated This method is deprecated since Flink 1.19. All metadata about the job should
+         *     be provided uniformly by {@link #getJobInfo()}.
+         * @see <a
+         *     href="https://cwiki.apache.org/confluence/display/FLINK/FLIP-382%3A+Unify+the+Provision+of+Diverse+Metadata+for+Context-like+APIs">
+         *     FLIP-382: Unify the Provision of Diverse Metadata for Context-like APIs </a>
          */
+        @Deprecated
```
Comment (xintongsong) [inline flink-core/src/main/java/org/apache/flink/core/failure/FailureEnricher.java:97]: Same here for the deprecated methods.

File: flink-core/src/main/java/org/apache/flink/core/failure/FailureEnricher.java [hunk 15]
```diff
@@ -124,5 +140,16 @@ enum FailureType {
...
         Executor getIOExecutor();
+
+        /**
+         * Get the meta information of current job.
+         *
+         * @return the job meta information.
+         */
+        @PublicEvolving
+        @Nullable
+        default JobInfo getJobInfo() {
```
Comment (xintongsong) [inline flink-core/src/main/java/org/apache/flink/core/failure/FailureEnricher.java:151]: Same here for default and nullable.

File: flink-core/src/main/java/org/apache/flink/api/common/JobInfoImpl.java (hunk 1 above)
Comment (WencongLiu) [reply flink-core/src/main/java/org/apache/flink/api/common/JobInfoImpl.java]: The `JobInfoImpl` class has been annotated by `@Internal`.

File: flink-core/src/main/java/org/apache/flink/api/common/TaskInfoImpl.java (hunk 2 above)
Comment (WencongLiu) [reply flink-core/src/main/java/org/apache/flink/api/common/TaskInfoImpl.java]: The `TaskInfoImpl` class has been annotated by `@Internal`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java (hunk 3 above)
Comment (WencongLiu) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java:173]: These code changes are prerequisite for the FLIP-380. I've moved these code changes to the commit `[FLINK-33905][core] Add getJobInfo() method to Environment`.

File: flink-core/src/main/java/org/apache/flink/api/common/functions/RuntimeContext.java (hunk 4 above)
Comment (WencongLiu) [reply flink-core/src/main/java/org/apache/flink/api/common/functions/RuntimeContext.java]: I've changed the default methods `getJobInfo()` and `getTaskInfo()` to be abstract.

File: flink-core/src/main/java/org/apache/flink/api/common/functions/RuntimeContext.java (hunk 5 above)
Comment (WencongLiu) [reply flink-core/src/main/java/org/apache/flink/api/common/functions/RuntimeContext.java]: Good point! For these deprecated methods in the modified interfaces, I've changed them from abstract to default and provide default implementations.

File: flink-core/src/test/java/org/apache/flink/api/common/operators/base/FlatMapOperatorCollectionTest.java (hunk 6 above)
Comment (WencongLiu) [reply flink-core/src/test/java/org/apache/flink/api/common/operators/base/FlatMapOperatorCollectionTest.java:46]: These code-style clean-ups have been placed into the commit `[hotfix] Fix the lacking class comments for the tests using TaskInfo`.

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/operator/CepRuntimeContextTest.java (hunk 7 above)
Comment (WencongLiu) [reply flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/operator/CepRuntimeContextTest.java:117]: These tests are conflicted with the newly added correctness check in the `TaskInfoImpl`. For example, the subtask ID shouldn't be large than or equal to the operator parallelism number. I've fixed these errors in the commit `[hotfix] Fix the logical errors for the tests related to task metadata`.

File: flink-connectors/flink-connector-files/src/test/java/org/apache/flink/connector/file/table/batch/compact/BatchPartitionCommitterSinkTest.java (hunk 8 above)
Comment (WencongLiu) [reply flink-connectors/flink-connector-files/src/test/java/org/apache/flink/connector/file/table/batch/compact/BatchPartitionCommitterSinkTest.java:130]: These tests are conflicted with the newly added correctness check in the `TaskInfoImpl`. For example, the subtask ID shouldn't be large than or equal to the operator parallelism number. I've fixed these errors in the commit `[hotfix] Fix the logical errors for the tests related to task metadata`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/execution/Environment.java (hunk 9 above)
Comment (WencongLiu) [reply flink-runtime/src/main/java/org/apache/flink/runtime/execution/Environment.java:127]: These code changes are prerequisite for the FLIP-380. I've moved these code changes to the commit `[FLINK-33905][core] Add getJobInfo() method to Environment`.

File: flink-core/src/main/java/org/apache/flink/api/connector/source/SourceReaderContext.java (hunk 10 above)
Comment (WencongLiu) [reply flink-core/src/main/java/org/apache/flink/api/connector/source/SourceReaderContext.java]: For these deprecated methods in the modified interfaces, I've changed them from abstract to default and provide default implementations.

File: flink-core/src/main/java/org/apache/flink/api/connector/source/SourceReaderContext.java (hunk 11 above)
Comment (WencongLiu) [reply flink-core/src/main/java/org/apache/flink/api/connector/source/SourceReaderContext.java]: I've changed the default methods including `getJobInfo()` and `getTaskInfo()` to be abstract.

File: flink-core/src/main/java/org/apache/flink/api/connector/sink2/InitContext.java (hunk 12 above)
Comment (WencongLiu) [reply flink-core/src/main/java/org/apache/flink/api/connector/sink2/InitContext.java]: I've changed the default methods including `getJobInfo()` and `getTaskInfo()` to be abstract.

File: flink-core/src/main/java/org/apache/flink/api/connector/sink2/InitContext.java (hunk 13 above)
Comment (WencongLiu) [reply flink-core/src/main/java/org/apache/flink/api/connector/sink2/InitContext.java]: I have annotated these deprecated methods by `@Deprecated` in the `InitContext`.

File: flink-core/src/main/java/org/apache/flink/core/failure/FailureEnricher.java (hunk 14 above)
Comment (WencongLiu) [reply flink-core/src/main/java/org/apache/flink/core/failure/FailureEnricher.java:95]: For these deprecated methods in the modified interfaces, I've changed them from abstract to default and provide default implementations.

File: flink-core/src/main/java/org/apache/flink/core/failure/FailureEnricher.java (hunk 15 above)
Comment (WencongLiu) [reply flink-core/src/main/java/org/apache/flink/core/failure/FailureEnricher.java]: I've changed the default methods including `getJobInfo()` and `getTaskInfo()` to be abstract.

File: pom.xml [hunk 16]
```diff
@@ -2301,6 +2301,8 @@ under the License.
 								<exclude>org.apache.flink.streaming.api.environment.StreamExecutionEnvironment#fromData(java.lang.Class,java.lang.Object[])</exclude>
 								<exclude>org.apache.flink.streaming.api.environment.StreamExecutionEnvironment#fromData(java.util.Collection)</exclude>
 								<exclude>org.apache.flink.streaming.api.environment.StreamExecutionEnvironment#fromData(java.util.Collection,org.apache.flink.api.common.typeinfo.TypeInformation)</exclude>
+								<exclude>org.apache.flink.api.common.functions.RuntimeContext</exclude>
+								<exclude>org.apache.flink.api.connector.source.SourceReaderContext</exclude>
```
Comment (xintongsong) [inline pom.xml:2305]: This should not be a hotfix commit. These change does not make sense by itself without the subsequent commits.

Comment (WencongLiu) [reply pom.xml:2305]: I have merged these changes to the commit where it should be.
